cnchi/mirrors.py

```diff
@@ -30,45 +30,324 @@
+                self.mirrors.append((url, active))
+            except KeyError:
+                pass
+
+    def fillme(self):
+        """ Fill listbox with mirrors info """
+        for listboxrow in self.get_children():
+            listboxrow.destroy()
+
+        drag_cbs = {
+            'drag-begin': self.on_drag_begin,
+            'drag-data-get': self.on_drag_data_get,
+            'drag-data-received': self.on_drag_data_received
+        }
+
+        for (url, active) in self.mirrors:
+            box = Gtk.Box(spacing=20)
+            box.set_name(url)
+            row = MirrorListBoxRow(
+                url,
+                active,
+                self.on_switch_activated,
+                drag_cbs)
+            self.add(row)
+        self.show_all()
 
-        self.disable_rank_mirrors = params["disable_rank_mirrors"]
+    def set_mirror_active(self, url, active):
+        """ Changes the active status in our mirrors list """
+        for index, item in enumerate(self.mirrors):
+            (murl, mact) = item
+            if url == murl:
+                self.mirrors[index] = (url, active)
+
+    def get_active_mirrors(self):
+        """ Returns a list with all active mirrors """
+        active_mirrors = []
+        for (url, active) in self.mirrors:
+            if active:
+                active_mirrors.append(url)
+        return active_mirrors
 
-        self.listbox_rows = {}
+    def on_switch_activated(self, switch, gparam):
+        row = switch.get_ancestor(Gtk.ListBoxRow)
+        if row:
+            self.set_mirror_active(row.data, switch.get_active())
+            self.emit("switch-activated")
+
+    def on_drag_begin(self, widget, drag_context):
+        """ User starts a drag """
+        row = widget.get_ancestor(Gtk.ListBoxRow)
+        alloc = row.get_allocation()
+        surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, alloc.width, alloc.height)
+        ctx = cairo.Context(surface)
+
+        row.get_style_context().add_class("drag-icon")
+        row.draw(ctx)
+        row.get_style_context().remove_class("drag-icon")
+
+        (x, y) = widget.translate_coordinates (row, 0, 0)
+
+        surface.set_device_offset(-x, -y)
+        Gtk.drag_set_icon_surface(drag_context, surface)
+
+    def on_drag_data_get(self, widget, drag_context, selection_data, info, time):
+        """ When drag data is requested by the destination """
+        row = widget.get_ancestor(Gtk.ListBoxRow)
+        listbox_str = str(self)
+        row_index = row.get_index()
+        data = "{0}|{1}".format(listbox_str, row_index)
+        selection_data.set_text(data, len(data))
+
+    def on_drag_data_received(self, widget, drag_context, x, y, selection_data, info, time):
+        """ When drag data is received by the destination """
+        data = selection_data.get_text()
+        try:
+            listbox_str = data.split('|')[0]
+            if listbox_str == str(self):
+                row_index = widget.get_index()
```

Comment:
__[flake8]__
* [F841] local variable 'row_index' is assigned to but never used

<!-- Comment by [SideCI](https://sideci.com) -->